yt_dlp/extractor/tiktok.py

```diff
@@ -108,9 +114,171 @@ def _extract_aweme(self, props_data, webpage, url):
+            'is_my_cn': 0,
+            'aid': 1180,
+            'ssmix': 'a',
+            'as': 'a1qwert123',
+            'cp': 'cbfhckdckkde1',
+        }
+
+        self._set_cookie('.tiktokv.com', 'odin_tt', ''.join(random.choice('0123456789abcdef') for i in range(160)))
+
+        aweme_detail = self._download_json(
+            'https://api-t2.tiktokv.com/aweme/v1/aweme/detail/', aweme_id,
+            'Downloading video details', 'Unable to download video details',
+            headers={
+                'User-Agent': f'com.ss.android.ugc.trill/{self._MANIFEST_APP_VERSION} (Linux; U; Android 10; en_US; Pixel 4; Build/QQ3A.200805.001; Cronet/58.0.2991.0)',
+            }, query=query)['aweme_detail']
+
+        video_info = aweme_detail['video']
+        formats = []
+
+        def extract_addr(addr, add_meta={}):
+            addr_formats = []
+            for url in addr['url_list']:
+                addr_formats.append({
+                    'url': url,
+                    'filesize': int_or_none(addr.get('data_size')),
+                    'ext': 'mp4',
+                    'acodec': 'aac',
+                    **add_meta,
+                })
+            return addr_formats
+
+        # Hack: Add direct video links first to prioritize them when removing duplicate formats
+        if video_info.get('play_addr'):
+            formats.extend(extract_addr(video_info['play_addr'], {
+                'format_id': 'play_addr',
+                'format_note': 'Direct video',
+                'vcodec': 'h265' if traverse_obj(
+                    video_info, 'is_bytevc1', 'is_h265') else 'h264',  # Always h264?
+                'width': video_info.get('width'),
+                'height': video_info.get('height'),
+            }))
+        if video_info.get('play_addr_h264'):
+            formats.extend(extract_addr(video_info['play_addr_h264'], {
+                'format_id': 'play_addr_h264',
+                'format_note': 'Direct video',
+                'vcodec': 'h264',
+                'width': video_info.get('width'),
+                'height': video_info.get('height')
+            }))
+        if video_info.get('play_addr_bytevc1'):
+            formats.extend(extract_addr(video_info['play_addr_bytevc1'], {
+                'format_id': 'play_addr_bytevc1',
+                'format_note': 'Direct video',
+                'vcodec': 'h265',
+                'width': video_info.get('width'),
+                'height': video_info.get('height')
+            }))
+        if video_info.get('download_addr'):
+            formats.extend(extract_addr(video_info['download_addr'], {
+                'format_id': 'download_addr',
+                'format_note': 'Download video%s' % (', watermarked' if video_info.get('has_watermark') else ''),
+                'vcodec': 'h264',
+                'width': video_info.get('width'),
+                'height': video_info.get('height'),
+                'source_preference': -2,
```

Comment:
why?

Reply:
to prioritize the non-watermarked video URLs

Reply:
then it makes more sense to set `preference` since it should be given lower priority under any sorting

```suggestion
                'preference': -2 if video_info.get('has_watermark') else -1,
```

Reply:
> to prioritize the non-watermarked video URLs

I hope watermarked hevc is still picked as best, right?

Reply:
no. I assume non-watermarked should always be picked if possible. Should it not?

Reply:
nope
Not good for archiving.

Reply:
watermarked videos are assumed to be the norm. If anyone want's without it then they can use -f

Also non watermarked will be absent if vid is obtained from webpage